sahi/predict.py

```diff
@@ -460,14 +472,97 @@ def predict(
     # iterate over source images
     durations_in_seconds["prediction"] = 0
     durations_in_seconds["slice"] = 0
-    for ind, image_path in enumerate(tqdm(image_path_list, "Performing inference on images")):
+
+    # Input video
+    if export_video:
+        if video_path is not None:
+            #get video name with extension
+            image_base = os.path.basename(video_path)
+        else:
+            print("\nVideo path is none. Enter (--video_path 'path' --export_video) in terminal!\n")
+        # get video from video path
+        cap = cv2.VideoCapture(video_path)
+        # get video properties
+        fps = cap.get(cv2.CAP_PROP_FPS)
+        w = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
+        h = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
+        size = (w, h)
+
+        fourcc = cv2.VideoWriter_fourcc(*'mp4v')
+        out = cv2.VideoWriter(str(save_dir) + '/{}'.format(image_base), fourcc, fps, size)
+
+        if view_image:
+            cv2.namedWindow('Prediction of {}'.format(str(image_base)), cv2.WINDOW_AUTOSIZE)
+        
+        if forward_backward:
+            
+            def get_video(cap):
```

Comment:
Rather than defining a sub-function inside `predict`, move this function out (maybe to utils @fcakyon ?) with a parameter `fast_forwarding: bool`, and handle this case in the function.

Reply:
Did you decide that? @devrimcavusoglu @fcakyon

Reply:
@madenburak we can create a function in `sahi/utils/ct.py` named `get_video_reader` as:

```python
def get_video_reader(video_path: str, view_image: bool = False, fast_forwarding_backwarding: bool = False):
    # get video name with extension
    image_base = os.path.basename(video_path)
    # get video from video path
    cap = cv2.VideoCapture(video_path)

    if fast_forwarding_backwarding:
        def read_video_frames(cap):
            while cap.isOpened:
                k = cv2.waitKey(30)

                frame_num = cap.get(cv2.CAP_PROP_POS_FRAMES)

                if k == 27:
                    print(
                        "\n===========================Closing==========================="
                    )  # Exit the prediction, Key = Esc
                    exit()
                if k == 100:
                    frame_num += 100  # Skip 100 frames, Key = d
                if k == 97:
                    frame_num -= 100  # Prev 100 frames, Key = a
                if k == 102:
                    frame_num += 5  # Skip 5 frames, Key = g
                if k == 103:
                    frame_num -= 5  # Prev 5 frames, Key = f
                cap.set(cv2.CAP_PROP_POS_FRAMES, frame_num)

                ret, frame = cap.read()
                if not ret:
                    print("\n===========================Video Ended===========================")
                    break
                yield Image.fromarray(frame)

                if view_image:
                    cv2.imshow("Prediction of {}".format(str(image_base)), get_image_from_dict)
                    cv2.waitKey(1)

    else:
        def read_video_frames(cap):
            while cap.isOpened:
                k = cv2.waitKey(30)

                if k == 27:
                    print(
                        "\n===========================Closing==========================="
                    )  # Exit the prediction, Key = Esc
                    exit()

                ret, frame = cap.read()
                if not ret:
                    print("\n===========================Video Ended===========================")
                    break
                yield Image.fromarray(frame)

                if view_image:
                    cv2.imshow("Prediction of {}".format(str(image_base)), get_image_from_dict)
                    cv2.waitKey(1)

    return read_video_frames
```

Reply:
then you can use this `get_video_reader` in `sahi/predict.py` as:

```python
video_capture, read_video_frame = get_video_reader(video_path, view_image, fast_forwarding_backwarding)
item = read_video_frame(video_capture)
```

Reply:
Ok. I will edit this.